src/tms_service.ts

```diff
@@ -0,0 +1,296 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import _ from 'lodash';
+import { EMSClient, EmsTmsFormat, TMSServiceConfig } from './ems_client';
+import { AbstractEmsService } from './ems_service';
+import { Sources, Style, VectorSource } from 'mapbox-gl';
+
+type EmsVectorSource = VectorSource & {
+  url: string;
+  tiles: string[];
+};
+
+type EmsVectorSources = Sources & {
+  [sourceName: string]: EmsVectorSource;
+};
+
+type EmsVectorStyle = Style & {
+  sources: EmsVectorSources;
+  sprite: string;
+  glyphs: string;
+};
+
+type EmsSprite = {
+  height: number;
+  pixelRatio: number;
+  width: number;
+  x: number;
+  y: number;
+};
+
+type EmsSpriteSheet = {
+  [spriteName: string]: EmsSprite;
+};
+
+type EmsRasterStyle = {
+  tilejson: string;
+  name: string;
+  attribution: string;
+  minzoom: number;
+  maxzoom: number;
+  bounds: number[];
+  format: string;
+  type: string;
+  tiles: string[];
+  center: number[];
+};
+
+export class TMSService extends AbstractEmsService {
+  protected readonly _config: TMSServiceConfig;
+
+  private _getRasterStyleJson = _.once(
```

Comment:
note: `_getRasterStyleJson`, `_getVectorStyleJsonRaw` and `_getVectorStyleJsonInlined` appear before the constructor because of the way they are instantiated. The [member-ordering rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/member-ordering.md) fails otherwise. 

If readability is still a concern, we could try to refactor these private methods. Suggestions are welcome.

Reply:
No need, readability is good, and maybe something we can improve later, but for now I'd go ahead with the code as it is